File: web/src/lib/api/git.ts
import type { Method } from 'api-typify';

export interface GitRepository {
	id: string;
	name: string;
	description: string;
	public: boolean;
	url: string;
	updatedAt: string;
}

export interface GitRepositories {
	repos: GitRepository[];
	limit: number;
	limitExceeded: boolean;
}

export interface GitInstallation {
	id: string;
	name: string;
	avatarURL: string;
	isUser: boolean;
	repositoryScopes: 'all' | 'selected';
	type: 'github';
}

export interface GitInstallations {
	installations: GitInstallation[];
	isLastPage: boolean;
}

type GET = Method<{
	'/v1/git/github/installations': {
		res: GitInstallations;
		queries?: {
			page?: number;
			perPage?: number;
		};
	};
	'/v1/git/github/installations/{installationID}/repos': {
		res: GitRepositories;
		queries?: {
			query?: string;
			name?: string;
			owner?: string;
		};
	};
}>;

export interface GitAPI {
	GET: GET;
}

File: types/http/git.go
package typesHTTP

import (
	"time"

	"github.com/alfiejones/panda-ci/types"
)

type GitRepository struct {
	ID          string    `json:"id"`
	Name        string    `json:"name"`
	UpdatedAt   time.Time `json:"updatedAt"`
	Description string    `json:"description"`
	Public      bool      `json:"public"`
	URL         string    `json:"url"`
}

type GitRepositories struct {
	Repos         []GitRepository `json:"repos"`
	LimitExceeded bool            `json:"limitExceeded"`
	Limit         int             `json:"limit"`
}

type GitInstallation struct {
	ID               string                `json:"id"`
	Name             string                `json:"name"`
	AvatarURL        string                `json:"avatarURL"`
	IsUser           bool                  `json:"isUser"`
	RepositoryScopes string                `json:"repositoryScopes"`
	AccountID        string                `json:"accountID"`
	Type             types.GitProviderType `json:"type"`
}

type GitInstallations struct {
	Installations []GitInstallation `json:"installations"`
	IsLastPage    bool              `json:"isLastPage"`
}
